client/src/main/java/io/pravega/client/stream/impl/WatermarkReaderImpl.java

```diff
@@ -0,0 +1,209 @@
+            }
+            iter = client.readFrom(location);
+            if (iter.hasNext()) {
+                Entry<Revision, Watermark> next = iter.next();
+                location = next.getKey();
+                return next.getValue();
+            }
+            return null;
+        }
+    }
+
+    private final Consumer<Map<SegmentWithRange, Long>> processFunction = new Consumer<Map<SegmentWithRange, Long>>() {
+
+        @Override
+        public void accept(Map<SegmentWithRange, Long> position) {
+            synchronized (lock) {
+                while (!inflight.isEmpty()) {
+                    int compare = compare(stream, position, inflight.getFirst());
+                    if (compare > 0) {
+                        passedTimestamp = inflight.removeFirst().getLowerTimeBound();
+                    } else {
+                        break;
+                    }
+                }
+                while (inflight.isEmpty() || compare(stream, position, inflight.getLast()) >= 0) {
+                    Watermark mark = fetcher.fetchNextMark();
+                    if (mark == null) {
+                        break;
+                    }
+                    if (compare(stream, position, mark) <= 0) {
+                        inflight.addLast(mark);
+                    } else {
+                        passedTimestamp = mark.getLowerTimeBound();
+                    }
+                }
+            }
+        }
+    };
+
+    /**
+     * Creates a Reader to keep track of the current time window for a given stream.
+     * 
+     * @param stream The stream to read the water marks for.
+     * @param clientFactory A factory to create the water mark reader. NOTE: This must use the same scope as the `stream` passed.
+     * @param executor A threadpool to perform background operations.
+     */
+    public WatermarkReaderImpl(Stream stream, SynchronizerClientFactory clientFactory, Executor executor) {
+        this.stream = Preconditions.checkNotNull(stream);       
+        
+        SynchronizerConfig config = SynchronizerConfig.builder().readBufferSize(4096).build();
+        this.fetcher = new WatermarkFetcher(clientFactory.createRevisionedStreamClient(StreamSegmentNameUtils.getMarkForStream(stream.getStreamName()),
+                                                                                       new WatermarkSerializer(),
+                                                                                       config));
+        this.processor = new LatestItemSequentialProcessor<>(processFunction, Preconditions.checkNotNull(executor));     
+    }
+    
+    /**
+     * Advances the position in the watermark segment based on the provided position.
+     * 
+     * @param readerGroupPosition - The latest read positions of the open segments in the reader group.
+     */
+    public void advanceTo(Map<SegmentWithRange, Long> readerGroupPosition) {
+        processor.updateItem(readerGroupPosition);      
```

Comment:
If `WatermarkReaderImpl.advanceTo(...)` is invoked when  `processor.updateItem(..)` is being processed then it will be ignored. Could you explain the scenario when that would happen ?

Reply:
If multiple calls to advanceTo happen in rapid succession this logic ensures that the last one will get processed but not all of the intermediate updates will necessarily be processed depending on how long it takes to do the processing. This is OK because a subsequent update is going to be a strictly greater position as the readers only move forward. It's a good idea because it ensures that the logic to track the marks can't fall behind, and doesn't end up wasting cycles trying to 'catch up' when it can just skip ahead.